// @elizaos/cli/dist/utils-QKNGJZLP.js

import { createRequire } from 'module';
const require = createRequire(import.meta.url);

import {
  NAV_BACK,
  NAV_NEXT,
  TestRunner,
  UserEnvironment,
  branchExists,
  buildGitHubSpecifier,
  buildProject,
  checkAndShowUpdateNotification,
  checkPluginRequirements,
  checkServer,
  configSchema,
  configureDatabaseSettings,
  confirmAction,
  copyClientDist,
  copyDir,
  copyTemplate,
  createBranch,
  createGitHubRepository,
  createPullRequest,
  detectPluginContext,
  displayAgent,
  displayBanner,
  ensureDirectory,
  ensureElizaDir,
  ensurePluginBuilt,
  executeInstallation,
  executeInstallationWithFallback,
  expandTildePath,
  findNextAvailablePort,
  forkExists,
  forkRepository,
  formatEnvFileWithTemplate,
  getAuthenticatedUser,
  getCliInstallTag,
  getConfigFilePath,
  getElizaDirectories,
  getEnvFilePath,
  getFileContent,
  getGitHubCredentials,
  getInstallCommand,
  getLatestCliVersion,
  getLocalPackages,
  getMissingEnvVars,
  getPackageInfo,
  getPackageManager,
  getPackageVersion,
  getPluginStatus,
  getVersion,
  handleError,
  installPlugin,
  isBunAvailable,
  isCliInstalledViaNpm,
  isGlobalInstallation,
  isPortFree,
  isRunningViaBunx,
  isRunningViaNpx,
  isUtf8Locale,
  isValidAnthropicKey,
  isValidGoogleKey,
  isValidOllamaEndpoint,
  isValidOpenAIKey,
  isValidOpenRouterKey,
  isValidPostgresUrl,
  loadConfig,
  loadEnvironment,
  loadPluginModule,
  logHeader,
  mergeProcessEnvWithTemplate,
  migrateCliToBun,
  promptAndStoreAnthropicKey,
  promptAndStoreGoogleKey,
  promptAndStoreOllamaConfig,
  promptAndStoreOllamaEmbeddingConfig,
  promptAndStoreOpenAIKey,
  promptAndStoreOpenRouterKey,
  promptAndStorePostgresUrl,
  promptForEnvVars,
  promptForMultipleItems,
  promptWithNav,
  provideLocalPluginGuidance,
  publishToGitHub,
  publishToNpm,
  pushToGitHub,
  rawConfigSchema,
  readEnvFile,
  removeFromBunLock,
  resolveConfigPaths,
  resolveEnvFile,
  resolveImport,
  resolvePgliteDir,
  saveConfig,
  saveGitHubCredentials,
  setupEnvFile,
  setupPgLite,
  showUpdateNotification,
  storeAnthropicKey,
  storeGoogleKey,
  storeOllamaConfig,
  storeOpenAIKey,
  storeOpenRouterKey,
  storePgliteDataDir,
  storePostgresUrl,
  testPublishToGitHub,
  testPublishToNpm,
  updateFile,
  validateEnvVars,
  validateGitHubToken,
  validatePluginEnvVars,
  writeEnvFile
} from "./chunk-IYFVOAOM.js";
import {
  runBunCommand
} from "./chunk-RIAWNDYI.js";
import "./chunk-KB3JDWUI.js";
import "./chunk-F24MS2YR.js";
import "./chunk-AQ6OMR2A.js";
export {
  NAV_BACK,
  NAV_NEXT,
  TestRunner,
  UserEnvironment,
  branchExists,
  buildGitHubSpecifier,
  buildProject,
  checkAndShowUpdateNotification,
  checkPluginRequirements,
  checkServer,
  configSchema,
  configureDatabaseSettings,
  confirmAction,
  copyClientDist,
  copyDir,
  copyTemplate,
  createBranch,
  createGitHubRepository,
  createPullRequest,
  detectPluginContext,
  displayAgent,
  displayBanner,
  ensureDirectory,
  ensureElizaDir,
  ensurePluginBuilt,
  executeInstallation,
  executeInstallationWithFallback,
  expandTildePath,
  findNextAvailablePort,
  forkExists,
  forkRepository,
  formatEnvFileWithTemplate,
  getAuthenticatedUser,
  getCliInstallTag,
  getConfigFilePath,
  getElizaDirectories,
  getEnvFilePath,
  getFileContent,
  getGitHubCredentials,
  getInstallCommand,
  getLatestCliVersion,
  getLocalPackages,
  getMissingEnvVars,
  getPackageInfo,
  getPackageManager,
  getPackageVersion,
  getPluginStatus,
  getVersion,
  handleError,
  installPlugin,
  isBunAvailable,
  isCliInstalledViaNpm,
  isGlobalInstallation,
  isPortFree,
  isRunningViaBunx,
  isRunningViaNpx,
  isUtf8Locale,
  isValidAnthropicKey,
  isValidGoogleKey,
  isValidOllamaEndpoint,
  isValidOpenAIKey,
  isValidOpenRouterKey,
  isValidPostgresUrl,
  loadConfig,
  loadEnvironment,
  loadPluginModule,
  logHeader,
  mergeProcessEnvWithTemplate,
  migrateCliToBun,
  promptAndStoreAnthropicKey,
  promptAndStoreGoogleKey,
  promptAndStoreOllamaConfig,
  promptAndStoreOllamaEmbeddingConfig,
  promptAndStoreOpenAIKey,
  promptAndStoreOpenRouterKey,
  promptAndStorePostgresUrl,
  promptForEnvVars,
  promptForMultipleItems,
  promptWithNav,
  provideLocalPluginGuidance,
  publishToGitHub,
  publishToNpm,
  pushToGitHub,
  rawConfigSchema,
  readEnvFile,
  removeFromBunLock,
  resolveConfigPaths,
  resolveEnvFile,
  resolveImport,
  resolvePgliteDir,
  runBunCommand,
  saveConfig,
  saveGitHubCredentials,
  setupEnvFile,
  setupPgLite,
  showUpdateNotification,
  storeAnthropicKey,
  storeGoogleKey,
  storeOllamaConfig,
  storeOpenAIKey,
  storeOpenRouterKey,
  storePgliteDataDir,
  storePostgresUrl,
  testPublishToGitHub,
  testPublishToNpm,
  updateFile,
  validateEnvVars,
  validateGitHubToken,
  validatePluginEnvVars,
  writeEnvFile
};
